Add X-Forwarded-For proxy header

mikejhill opened the pull request:
Targeting this fork since it seems to be the most active of UrlRewriteFilter and has been a really helpful fork for us so far (thank you for your work!).

This tries to fix an issue into where (reverse-)proxying doesn't include the X-Forwarded-For header. This header propagates the original client IP to downstream targets. This need for us comes out of a simple (Atlassian Crowd) SSO solution that requires validating an SSO token against a matching client IP. Hosts need to see the real client IP during proxied requests rather than just the proxy (UrlRewriteFilter) server's IP.

MDN X-Forwarded-For page: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For

There seems to be some question on whether X-Forwarded-For or X-Real-IP is a better header to use ([source](https://github.com/akka/akka-http/issues/1670#issuecomment-353072461)). However, MDN doesn't list X-Real-IP, and I don't seem to see it used as often in other places. The internal NGINX reverse proxy configurations that I use elsewhere set X-Real-IP and X-Forwarded-{For,By,Proto} (can also optionally add X-Forwarded-{Host,Port,Server}), but the only one that is actually used by our Tomcat instances is X-Forwarded-For. NGINX mentions X-Real-IP in some of their guides, but seems to prefer X-Forwarded-For (e.g., `$proxy_add_x_forwarded_for`). The Tomcat [RemoteIpValve](https://tomcat.apache.org/tomcat-8.5-doc/api/org/apache/catalina/valves/RemoteIpValve.html) can also make use of a few other headers like X-Forwarded-By if desired. I don't see any mention of X-Real-IP in the org.apache.tomcat:tomcat-catalina:9.0.30 source.

machak commented:
@mikejhill thanks for the patch, I'll review it tomorrow. 
I believe we use X-Forwarded-For header in our stack (we can always add an option to make it configurable)

machak commented on src/main/java/org/tuckey/web/filters/urlrewrite/RequestProxy.java:
```diff
@@ -279,6 +288,22 @@ private static HttpRequestBase setupProxyRequest(final HttpServletRequest hsRequ
+            } else {
+                final String clientAddr = xForwardedForValue.split(",\\s*")[0].trim();
+                if (!clientAddr.isEmpty()) {
+                    xForwardedForValue += ", " + clientAddr;
```
Ok, I am bit puzzled by this part: do we need first address to also be added as last, if I follow code correctly,
foo, bar input will produce foo, bar, foo  output. Is last foo needed? 
As far as I know, first address listed is remote client address (foo in above case)

mikejhill replied:
Sorry, I bamboozled this part with some last-minute refactoring yesterday. The expected X-Forwarded-For format is "\<Client IP>, <Proxy 1 IP>, <Proxy 2 IP>, etc.", which was definitely not what I had here. This should be fixed as of fc7ff7572fd27c9f6e531d3b13fb22f3a8cade77.

boudekerk commented:
The options's called addProxyHeader**s**, so should X-Forwarded-Proto and X-Forwarded-Host then be forwarded/appended as well?

mikejhill commented:
> The options's called addProxyHeader**s**, so should X-Forwarded-Proto and X-Forwarded-Host then be forwarded/appended as well?

I've added additional options for X-Forwarded-Proto, X-Forwarded-Host, X-Forwarded-By, and Host (for manually propagating the host value) now. I also changed the `add-proxy-headers` DTD to accept a comma-delimited list of which of these headers to use.

I found myself needing differing headers depending on my configuration. For example, `RemoteIpValve` in Tomcat 9.0.20 doesn't support `X-Forwarded-Host`, so `Host` must be propagated manually. Tomcat 9.0.23 added support (https://bz.apache.org/bugzilla/show_bug.cgi?id=57665#c24), so in those cases we would prefer to use `X-Forwarded-Host` rather than change the real host header.

mikejhill commented:
Made some minor updates, added `X-Forwarded-Prefix` support, and added simple tests. I believe existing comments are resolved.
